Assets/Scripts/Noise/PerlinNoise.cs

```diff
@@ -0,0 +1,210 @@
+        // The left bound is ( |_x_|,|_y_|,|_z_| ) and the right bound is that
+        // plus 1.  Next we calculate the location (from 0.0 to 1.0) in that cube.
+        int xi = (int)x & 255;
+        int yi = (int)y & 255;
+        int zi = (int)z & 255;
+        float xf = x - (int)x;
+        float yf = y - (int)y;
+        float zf = z - (int)z;
+
+        // Compute faded values for smoothing
+        float u = Fade(xf);
+        float v = Fade(yf);
+        float w = Fade(zf);
+
+        // Perlin Noise hash function is used to get a unique value for every coordinate input.
+        // This is the hash for each of the 8 points on the unit cube around this point
+        int aaa, aba, aab, abb, baa, bba, bab, bbb;
+        aaa = p[p[p[    xi ]+    yi ]+    zi ];
+        aba = p[p[p[    xi ]+Inc(yi)]+    zi ];
+        aab = p[p[p[    xi ]+    yi ]+Inc(zi)];
+        abb = p[p[p[    xi ]+Inc(yi)]+Inc(zi)];
+        baa = p[p[p[Inc(xi)]+    yi ]+    zi ];
+        bba = p[p[p[Inc(xi)]+Inc(yi)]+    zi ];
+        bab = p[p[p[Inc(xi)]+    yi ]+Inc(zi)];
+        bbb = p[p[p[Inc(xi)]+Inc(yi)]+Inc(zi)];
+
+        float x1, x2, y1, y2;
+        // The gradient function calculates the dot product between a pseudorandom
+        // gradient vector and the vector from the input coordinate to the 8
+        // surrounding points in its unit cube.
+        // This is all then lerped together as a sort of weighted average based on the faded (u,v,w)
+        x1 = Lerp(Grad (aaa, xf  , yf  , zf),           
+                  Grad (baa, xf-1, yf  , zf),             
+                  u);                                     
+        x2 = Lerp(Grad (aba, xf  , yf-1, zf),           
+                  Grad (bba, xf-1, yf-1, zf),             
+                  u);
+        y1 = Lerp(x1, x2, v);
+
+        x1 = Lerp(Grad (aab, xf  , yf  , zf-1),
+                  Grad (bab, xf-1, yf  , zf-1),
+                  u);
+        x2 = Lerp(Grad (abb, xf  , yf-1, zf-1),
+                  Grad (bbb, xf-1, yf-1, zf-1),
+                  u);
+        y2 = Lerp (x1, x2, v);
+
+        // For convenience we bind the result to 0 - 1 (theoretical min/max before is [-1, 1])
+        return (Lerp (y1, y2, w)+1)/2;
+    }
+
+    /// <summary>
+    /// Computes multiple octave combination of perlin noise
+    /// </summary>
+    /// <param name="x">X value to use in coordinate space</param>
+    /// <param name="y">y value to use in coordinate space</param>
+    /// <param name="z">z value to use in coordinate space</param>
+    /// <param name="octaves">Number of octaves to apply</param>
+    /// <param name="persistence">Change in amplitude over each octave (decay)</param>
+    /// <param name="frequencyGrowth">Growth in frequency over each octave (growth factor)</param>
+    /// <returns></returns>
+    public float OctavePerlin(float x, float y, float z, int octaves, float persistence, float frequencyGrowth) {
```

Comment:
Another place that could take in a vector. It will help reduce the parameter length here

Reply:
Done, changed to use a vector.